Move java sdk configuration documentation into docs site (#4377)
open-telemetry/opentelemetry.io: content/en/docs/languages/java/configuration.md

{{% alert title="System Properties and Environment Variables" color="info" %}}
Any setting configurable with a system property can also be configured with an
environment variable. Apply the following steps to convert a system property to
an environment variable:

- Convert the name to uppercase.
- Replace all `.` and `-` characters with `_`.

For example `otel.instrumentation.common.default-enabled` would convert to
`OTEL_INSTRUMENTATION_COMMON_DEFAULT_ENABLED`.

{{% /alert %}}

## General

The autoconfigure module registers Java shutdown hooks to shut down the SDK when
appropriate. Please note that since this project uses `java.util.logging` for
all of it's logging, some of that logging may be suppressed during shutdown
hooks. This is a bug in the JDK itself, and not something we can control. If you
require logging during shutdown hooks, please consider using `System.out` rather
than a logging framework that might shut itself down in a shutdown hook, thus
suppressing your log messages. See this
[JDK bug](https://bugs.openjdk.java.net/browse/JDK-8161253) for more details.

> 

Reviewer: I recommend putting in another note about how signal refers to traces, metrics, logs:

```
{{% alert title="Signal configuration" color="info" %}}

The text  placeholder`{signal}` refers to the supported [OpenTelemetry Signal](/docs/concepts/signals/).
Valid values include `traces`, `metrics`, and `logs`.

{{% /alert %}}
```

That way folks don't think you need to specify it as `otel.{traces}.exporter` and the like.

---
title: Configuration
linkTitle: Configuration
weight: 10
aliases: [config]
# prettier-ignore
cSpell:ignore: authservice blrp Dotel ignore LOWMEMORY myservice ottrace PKCS retryable
---

The OpenTelemetry SDK provides a working implementation of the API, and can be
set up and configured in a number of ways. The Java SDK supports most of the
available [configuration options](/docs/languages/sdk-configuration/). For
conformance details, see the
[compliance matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md).

{{% alert title="System Properties and Environment Variables" color="info" %}}
Any setting configurable with a system property can also be configured with an
environment variable. Apply the following steps to convert a system property to
an environment variable:

- Convert the name to uppercase.
- Replace all `.` and `-` characters with `_`.

> For example `otel.instrumentation.common.default-enabled` would convert to
> `OTEL_INSTRUMENTATION_COMMON_DEFAULT_ENABLED`.

Reviewer 2: ```suggestion
For example `otel.sdk.enabled` would convert to
`OTEL_SDK_ENABLED`.
```

{{% /alert %}}

## General

> The autoconfigure module registers Java shutdown hooks to shut down the SDK when

Reviewer 2: Let's link to or provide some context on what the autoconfigure module actually is. We have to assume the reader doesn't know what the autoconfigure module is when they land on this page. Maybe a link to this: http://localhost:1313/docs/languages/java/instrumentation/#automatic-configuration

And that section should probably link back to this page for a full list of all options.

appropriate. Please note that since this project uses `java.util.logging` for
all of it's logging, some of that logging may be suppressed during shutdown
hooks. This is a bug in the JDK itself, and not something we can control. If you
require logging during shutdown hooks, please consider using `System.out` rather
than a logging framework that might shut itself down in a shutdown hook, thus
suppressing your log messages. See this
[JDK bug](https://bugs.openjdk.java.net/browse/JDK-8161253) for more details.

{{% alert title="Signal configuration" color="info" %}}

The text placeholder `{signal}` refers to the supported
[OpenTelemetry Signal](/docs/concepts/signals/). Valid values include `traces`,
`metrics`, and `logs`.

{{% /alert %}}

### Configuration Priority

Signal specific configurations take priority over the generic versions.

For example, if you set both `otel.exporter.otlp.endpoint` and

> `otel.exporter.otlp.traces.endpoint`, the latter will take precedence.

Reviewer 2: Should we just include this text about priority inside the "Signal configuration" alert?

Exporters output the telemetry. The following configuration properties are
common to all exporters:

| System property                               | Purpose                                                                                                                                                                                               |
| --------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `otel.{signal}.exporter`                      | List of exporters to be used for {signal}, separated by commas. Default is `otlp`. `none` means no auto-configured exporter.                                                                          |
| `otel.java.experimental.exporter.memory_mode` | If `reusable_data`, enable reusable memory mode (on exporters which support it) to reduce allocations. Default is `immutable_data`. This option is experimental and subject to change or removal.[^1] |

[^1]:

Exporters which adhere to
`otel.java.experimental.exporter.memory_mode=reusable_data` are

> `OtlpGrpcMetricExporter`, `OtlpHttpMetricExporter`, and `PrometheusHttpServer`.

Reviewer 2: ```suggestion
`OtlpGrpc{Signal}Exporter`, `OtlpHttp{Signal}Exporter`, and `PrometheusHttpServer`.
```
All the OTLP exporters now adhere to it.

Support for additional exporters may be added in the future.

> #### OTLP exporter (span, metric, and log exporters)

Reviewer 2: We should mention the artifact coordinates in each place where we describe configuration that requires a particular artifact to be included in order to work.

I think this comes up for each of the exporter sections (otlp, logging, logging-otlp, zipkin, prometheus), and for the propagators.

| `otel.experimental.exporter.otlp.retry.enabled`            | If `true`, enable [experimental retry support](#otlp-exporter-retry).                                                                                                                                                                                                                                                                                                                                                                | `false`                                                                                                                    |

[^2]:

> OpenTelemetry Java agent 2.x uses `http/protobuf` by default.

Reviewer 2: I think something is funky with the footnote display. When I run locally, clicking on the `2` footnote brings me to the bottom of the page.

> To configure the service name for the OTLP exporter, add the `service.name` key
> to the OpenTelemetry Resource. For example:
> `OTEL_RESOURCE_ATTRIBUTES=service.name=myservice`.

Reviewer 2: Let's get rid of this. And we might also consider bringing the resource section above exporters. 
```suggestion
```

##### OTLP exporter retry

[OTLP](/docs/specs/otlp/#otlpgrpc-response) requires that
[transient](/docs/specs/otel/protocol/exporter/#retry) errors be handled with a
retry strategy. When retry is enabled, retryable gRPC status codes are retried
using an exponential backoff with jitter algorithm as described in the
[gRPC Retry Design](https://github.com/grpc/proposal/blob/master/A6-client-retries.md#exponential-backoff).

> The policy has the following configuration, which there is currently no way to
> customize:

Reviewer 2: ```suggestion
The policy has the following configuration, which is can only be customized via programmatic configuration (see [customizing the OpenTelemetry SDK](#customizing-the-opentelemetry-sdk)):
```

{{% alert title="Note" color="info" %}} While the
`OtlpJsonLogging{Signal}Exporters` are stable, specifying their use via
`otel.{signal}.exporter=logging-otlp` is experimental and subject to change or
removal. {{% /alert %}}

### Resources

A resource is the immutable representation of the entity producing the
telemetry. See [Resource semantic conventions](/docs/specs/semconv/resource/)
for more details.

| System Property                            | Description                                                                                                 |
| ------------------------------------------ | ----------------------------------------------------------------------------------------------------------- |
| `otel.resource.attributes`                 | Specify resource attributes in the following format: `key1=val1,key2=val2,key3=val3`.                       |
| `otel.service.name`                        | Specify logical service name. Takes precedence over `service.name` defined with `otel.resource.attributes`. |
| `otel.experimental.resource.disabled-keys` | Specify resource attribute keys that are filtered.                                                          |

> You almost always want to specify the
> [`service.name`](/docs/specs/semconv/resource/#service) for your application. It
> corresponds to how you describe the application, for example `authservice` could
> be an application that authenticates requests. If not specified, SDK defaults
> the service name to `unknown_service:java`.

Reviewer 2: ```suggestion
You almost always want to specify `otel.service.name` to set the [`service.name`](/docs/specs/semconv/resource/#service) resource attribute, which represents the logical name of your service. If not specified, the SDK sets `service.name=unknown_service:java` by default.
```

optionally use an artifact that includes built-in ResourceProviders:

- [io.opentelemetry.instrumentation:opentelemetry-resources](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/resources)
  includes providers for a
  [predefined set of common resources](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources)
- [io.opentelemetry.contrib:opentelemetry-aws-resources](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/aws-resources)
  includes providers for
  [common AWS resources](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/aws-resources/src/main/java/io/opentelemetry/contrib/aws/resource)
- [io.opentelemetry.contrib:opentelemetry-gcp-resources](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/gcp-resources)
  includes providers for
  [common GCP resources](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/gcp-resources/src/main/java/io/opentelemetry/contrib/gcp/resource)

### Disabling automatic ResourceProviders

> If you are using the `ResourceProvider` SPI, which many instrumentation agent
> distributions include automatically, you can turn on or off one or more of them
> by using the following configuration items:

Reviewer 2: ```suggestion
Many instrumentation agent distributions automatically various `ResourceProvider` implementations. These can be turned on or off as follows:
```

attributes.

| System property                     | Description                                                                            | Default  |
| ----------------------------------- | -------------------------------------------------------------------------------------- | -------- |
| `otel.attribute.value.length.limit` | The maximum length of attribute values. Applies to spans and logs.                     | No limit |
| `otel.attribute.count.limit`        | The maximum number of attributes. Applies to spans, span events, span links, and logs. | `128`    |

### Propagators

The propagators determine which distributed tracing header formats are used, and
which baggage propagation header formats are used.

| System property    | Description                                                                      | Default                      |
| ------------------ | -------------------------------------------------------------------------------- | ---------------------------- |
| `otel.propagators` | The propagators to be used. Use a comma-separated list for multiple propagators. | `tracecontext,baggage` (W3C) |

Supported values are the following:

> | Value          | Description                                                                                                      |

Reviewer 2: Can we add a column to this table with the artifact that contains each of these?

- `tracecontext` lives in `opentelemetry-api`
- `baggage` lives in `opentelemetry-api`
- `b3` lives in `opentelemetry-extension-trace-propagators`
- `b3multi` lives in `opentelemetry-extension-trace-propagators`
- `jaeger` lives in `opentelemetry-extension-trace-propagators`
- `xray` lives in `io.opentelemetry.contrib:opentelemetry-aws-xray-propagator`
- `ottrace` lives in `opentelemetry-extension-trace-propagators`
